types/node/buffer.d.ts

```diff
@@ -910,7 +917,7 @@ declare module "buffer" {
              * @param [start=0] Where the new `Buffer` will start.
              * @param [end=buf.length] Where the new `Buffer` will end (not inclusive).
              */
-            slice(start?: number, end?: number): Buffer;
+            slice(start?: number, end?: number): Buffer & WithArrayBufferLike<ArrayBuffer>;
```

Comment:
> The `array`/`obj` argument issue can be addressed by using `this` types. For `slice()` and `subarray()` this can either be addressed by either making `Buffer` generic over `TArrayBuffer`, or changing the return types of `slice()` and `subarray()`. Making `Buffer` generic is the more reliable option, but would require reintroducing a `"typesVersions"` entry to the NodeJS types to handle the inconsistency. A less invasive and more backwards compatible fix would be a minor change to the return types of `slice()` and `subarray()` in `Buffer`:
> 
> ```ts
> interface WithArrayBufferLike<TArrayBuffer extends ArrayBufferLike> {
>     readonly buffer: TArrayBuffer;
> }
> interface Buffer extends Uint8Array {
>     ...
>     slice(start?: number, end?: number): Buffer & WithArrayBufferLike<ArrayBuffer>;
>     subarray(start?: number, end?: number): Buffer & WithArrayBufferLike<this["buffer"]>;
>     ...
> }
> ```

The behaviour of `Buffer.prototype.slice` differs from its TypedArray counterpart in that it's just an alias to `Buffer.prototype.subarray`; the returned Buffer is a view on the existing Buffer's array buffer, in contrast to `TypedArray.prototype.slice` which copies into a new ArrayBuffer.

This should therefore be `WithArrayBufferLike<this["buffer"]>`, providing the compiler doesn't complain about the divergent return types.